### DISCUSSION TITLE
StreamField Block UI Enhancements: Drag & Drop, Copy/Paste, Styling and Previews

### DISCUSSION THREAD

**Discussion by benenright** [wagtail/wagtail#12354] [Ideas]
Hi

We have been exploring how to best incorporate drag-and-drop and copy-paste functionality into the StreamField interface to make it easier to manage and edit content blocks.

This is also part of ongoing efforts to improve the ease of navigation within the StreamField form area. We aim to make blocks easier to scan and locate by enhancing visibility, with clearer block titles and a more intuitive collapsed block view. Additionally, we want to help users better differentiate between block types in the form, minimap, and block chooser.

We've created a prototype that brings these elements together in a revamped version of the StreamField UI.

### Related issues
https://github.com/wagtail/wagtail/issues/8339 - Copy and Paste
https://github.com/wagtail/wagtail/issues/1033 - Drag and Drop
https://github.com/wagtail/wagtail/issues/11997 - Block chooser previews
https://github.com/wagtail/wagtail/issues/10055 - Minimap refinements

Our plan is to begin development on these features in the next release (post-November), and we are sharing this now to gather feedback and input, as there are several interrelated components and interactions here. I'm sure these design will now evolve.

[A semi-functional design prototype is here.](https://benenright.github.io/wagtail/) Have a play (in Chrome if poss)

### Screenshots

![image 1](https://github.com/user-attachments/assets/fa6570c5-1078-40f2-95e0-91cb72d624ed)
![image 3](https://github.com/user-attachments/assets/e507d62c-9121-4bcb-abaa-9025f8a211a4)
![image 4](https://github.com/user-attachments/assets/b8d61a83-d5e0-47dc-ab90-185c8774fa7e)
![image 5](https://github.com/user-attachments/assets/9f033b11-b32f-4e42-9b32-b48da33be30e)
![image 6](https://github.com/user-attachments/assets/d7d951e7-05b6-4c73-8f82-04c506151c20)
![image 7](https://github.com/user-attachments/assets/37ed97b6-85a4-4b25-816d-192a32f76f1f)
![image 8](https://github.com/user-attachments/assets/e21cae3d-5345-4837-bb5c-a2cd1f130051)
![model-dialog-template-1](https://github.com/user-attachments/assets/8df2974f-32d1-429a-92f9-f07141e397f9)
![model-dialog-template-2](https://github.com/user-attachments/assets/508f428b-99a1-45fb-b983-01848859eb4f)
![model-dialog-template-3](https://github.com/user-attachments/assets/c7c18e71-9a3d-45e6-9d26-44b1c159e3cd)

### Some design notes

- There are various ui discrepancies here with the actual wagtail build - this is just because it's a prototype, I'll mention the things below that we are proposing changing...
- Potential for a more visible style for block headings (inline with thinking here https://github.com/wagtail/wagtail/issues/11997) to make them easier to scan on long pages with large blocks: brighter colour, and keeping the block title visible alongside the 'preview text' when collapsed
- New styles carried through to the block chooser. We could add 'Paste block' in here too. The styling is off in the prototype, but it could also sit alongside 'Split block' when you have triggered the chooser from within a rich text area
- Addition of a 'drag' icon grouped with the move up/down arrow' (I think the whole top area of the block / whole of the grey bg on collapsed block would be the hit area for dragging)
- A more tactile, draggable-looking collapsed block
- We'd envisage removing the 'drag-back' animation of the blocks when they are successfully placed in a different location (I wasn't able to achieve this in the prototype)
- Collapsing large, open blocks while dragging, to make it an easier thing to drag
- Replace the existing 'Duplicate (below)' button with 'Copy (to clipboard)' (with toast notification to confirm). How do people feel about this change?
- A 'Paste block' button could be positioned next to the '+' button between blocks. We haven't yet explored the option to 'paste over' an existing block. Due to the technical nature of how Copy and Paste work, the Paste button needs to be consistently available. This is because Wagtail can't detect if there's a block on the clipboard—especially if the user copied content from a different StreamField area, page, or window—until the paste action is initiated. So we'd need to assess how intrusive this feels to the UI or if there are any other workarounds
- We have only started to explore messaging for when a paste is unsuccessful or compromised due to things like incompatible formatting between one SF and another (more on this here: https://github.com/wagtail/wagtail/issues/8339)
- The minimap in this prototype is incomplete. We hope to explore this later down the line as there are obvious parallels with things like block Drag and Drop between the form and the minimap - we want to consider if that functionality should also be added to the minimap

**Comment by benenright** [comment]
Exploring 'paste over' a block (feels clunky having this label) / paste inline (to split a block):

<img width="1596" alt="Screenshot 2024-10-08 at 13 35 48" src="https://github.com/user-attachments/assets/b5458dfa-ed14-4256-b8fa-47037c0d7b0e">

**Reply by tbrlpld** [reply]
`CTRL-V` for paste would be lovely.

**Reply by tbrlpld** [reply]
What does "paste over" mean? "over" as in "above"? Wouldn't that just be paste below the previous?

**Reply by tbrlpld** [reply]
Oh, just seeing the comment. It means to replace the existing block? I think I agree with @allcaps , this might create more confusion than paste and delete.

**Comment by allcaps** [comment]
Nice work! It looks good. I like the drag handles, paste icons, and the paste block in the actions section. 👏

I'm not sure about the 'paste over' option. The potential for confusion doesn't seem to outweigh the extra click required to delete and paste.

I notice there are several copy icons to put a single block on the clipboard. Will there be a feature to "copy all blocks from this StreamField"?

**Comment by tbrlpld** [comment]
@benenright I think this is going in a great direction. Drag-and-drop and a focus on scanability should definitely help making work with stream field smoother. 

A couple of thoughts that came up when looking at it: 
* It seems like the blocks are getting more and more buttons. Copy, delete, up, down, drag-handle. I wonder if this could be cleaned up, similar to how most page actions are in a drop-down. Especially the up and down buttons are somewhat superfluous once drag-and-drop is in place. 
* I wonder if it would help scanablility to move the "add" and "paste" buttons to the right. This would mean only the icons and labels are on the left side. Might help when scanning for a label to not see the same buttons over and over. 
* With the "past block" action in the "add" menu, I wonder if we even need the separate "paste" button. It is an extra click, but all other block additions also need that extra click. Also, if possible it would be great to have the `CTRL-V` shortcut work when the mouse is hovering over the dead-space between the blocks (the row where currently the "add" an "paste" buttons are).

**Reply by tbrlpld** [reply]
Quick mock-up of hidden buttons and "Add" button on the right.

<img width="500" alt="streamfield-editor-hidden-buttons" src="https://github.com/user-attachments/assets/4e50fb86-da80-472f-a640-3c34ef988a43">

**Comment by BigWhale** [comment]
This looks slick! Drag'n'drop will make certain things much easier. I am a bit against hiding buttons behind the menu, but we are working on optimising our interface to reduce the number of mouse operations required to write and publish something.

Not sure if this is the place, but I would like to mention the ability to group certain items in the block and more importantly, hide them. For example, in Quote block only the first field is required, the rest are optional. It would be great if I could hide/collapse optional blocks somehow. This would make the UI a bit more streamlined and less cluttered.

**Reply by benenright** [reply]
> Not sure if this is the place, but I would like to mention the ability to group certain items in the block and more importantly, hide them. For example, in Quote block only the first field is required, the rest are optional. It would be great if I could hide/collapse optional blocks somehow. This would make the UI a bit more streamlined and less cluttered.

Hey @BigWhale I think this could be related to: https://github.com/wagtail/wagtail/issues/10227

**Comment by jgardiner68** [comment]
As long as there are options to lock down the layout (eg gallery in a fixed location, but freedom to change the order of images within the gallery), then I am fine.

**Comment by liamjohnston** [comment]
Love it.

**Comment by krummmy** [comment]
Thank you very much, I can't wait for this improvement to be released. Can we expect the same drag & drop logic to be applied to InlinePanel elements?

**Comment by KalobTaulien** [comment]
This is AWESOME! It looks modern and feels good to use. Extra kudos for a semi-functional demo page, too! 

My feedback is very minor, and similar to others:
* The up and down arrows seem unneeded if there's a drag-and-drop icon (I feel most people understand what that is these days) though I can see why we might want to keep them for legacy users. 
* "paste over" may cause more confusion than it's worth. A copy-paste-then-delete-the-old seems to fall in line with a user experience I'm personally used to. It's more work, but feels like the standard. 

What I personally like that may conflict with others opinions:
* The "+" and "paste block" icons on the left -- I love it! I'm a left to right reader, so it makes sense. 
* I love the additional ideas other people have added or asked about, but I'd rather see these changes happen sooner with updates down the road. 

Whether any of the above mentioned notes are taken or not, I think this is outstanding work and a HUGE step forward for the Wagtail and the users adding content. 🙌

**Comment by tomusher** [comment]
This looks like an excellent improvement!

I'm not sure if this feedback is best left as a separate issue, or if it's something you'd want to address as part of these changes, but the most common bit of feedback I see is about the '+' buttons on nested blocks. Users find it hard to know which + button to press to create a block where they want it, and I get confused sometimes too.

<img width="1040" alt="Screenshot 2024-11-20 at 10 58 45" src="https://github.com/user-attachments/assets/8acc5b9c-2208-4e84-b1f8-ba344add307d">

In this example (from the current design), we've got the top-level StreamField, a Section block which can have multiple Elements, and in some cases one of those Elements can have multiple items - so you end up with 3 + buttons at the end of a block with no clear indication which will do what (the low contrast on the dotted lines in dark mode doesn't help here)

<img width="936" alt="Screenshot 2024-11-20 at 11 00 31" src="https://github.com/user-attachments/assets/f3050352-1aad-4864-90ee-aca21967df20">
In this example from the new mockups, I can still see how it could be confusing figuring out which button does what, although offsetting the + button from the dotted line seems to help visually.

Perhaps having some indication of where a block will appear through an on-hover silhouette, or by replacing the + button with a 'skeleton' block at the right 'level' would help? Alternatively, labelling the plus button based on the field name could help, e.g. in my first example, 'Add to Elements', 'Add to Info Blocks'.

**Reply by thibaudcolas** [reply]
Thank you, this is the right place!

**Comment by nilaygit-10721** [comment]
Feedback Themes:
Streamlined UX:

Many users agreed on decluttering, especially by hiding unnecessary buttons or grouping actions.
Skepticism around "Paste over" due to potential confusion; a standard copy-paste-delete approach is preferred.
Proposed Enhancements:
Drag-and-Drop Functionality:

Simplifies reordering blocks.
Includes a tactile, draggable look for collapsed blocks.
Suggestion to remove the current drag-back animation post-drop for smoother interaction.
Copy/Paste Improvements:

Replace "Duplicate (below)" with "Copy (to clipboard)" and add a "Paste block" button.
Include feedback messaging for incompatible pastes.
Enable CTRL+V as a shortcut for pasting in the dead-space between blocks.
Styling Improvements:

Clearer block titles with brighter colors and visible headings in collapsed mode.
Improved visibility in the block chooser, potentially including inline previews and descriptions.
Button Placement and Interaction:

Explore moving "Add" and "Paste" buttons to the right for easier scanning.
Potential removal of up/down arrows as drag-and-drop may make them redundant.
Discussed grouping actions in a dropdown to declutter the interface.

**Comment by sulram** [comment]
@benenright nice work!
